### ISSUE TITLE
Possibility of option to handle requeued jobs for SLURM?

### ISSUE THREAD

**Issue by dctrud** [nextflow-io/nextflow#226]
Hi,

On our SLURM cluster we have a setup where jobs that are first suspended may then be re-queued to vacate a node and start again later.

This means that it's possible for a job to go from PENDING -> RUNNING -> SUSPENDED -> PENDING -> RUNNING->COMPLETE

When a job is active, nextflow will mark it failed when it goes back to PENDING state. It'd be nice to have an option somewhere (maybe on the SLURM executor directives) to allow this re-queue flow.

I'm not sure whether this is straightforward (i.e just an option so nextflow doesn't fail an active job when it drops to PENDING), or whether it's a lot more involved (i.e. nextflow jobs are structured so that they cannot be re-started from the top).

Cheers.

**Comment by dctrud** [comment]
In SLURM docs:

http://slurm.schedmd.com/scontrol.html

requeue job_list
    Requeue a running, suspended or finished Slurm batch job into pending state. The job_list argument is a comma separated list of job IDs.

**Comment by pditommaso** [comment]
I think I've understood what's the cause of this issue. Both suspended and pending status should be mapped correctly by nextflow, see [here](https://github.com/nextflow-io/nextflow/blob/master/src/main/groovy/nextflow/executor/SlurmExecutor.groovy#L137-L149). 

But but the pending status is not included in the list of statues in which a job is considered [active](https://github.com/nextflow-io/nextflow/blob/master/src/main/groovy/nextflow/executor/AbstractGridExecutor.groovy#L315-315), thus when when it change from *suspended* to *pending* it could be marked as *failed*.

**Comment by pditommaso** [comment]
I've uploaded a snapshot that should solve the problem. You can use specifying the variable `NXF_VER=0.22.2-SNAPSHOT` in your environment. 

If still does not fix the issue, please try to isolate the problem in a small use case and run it enabling the nextflow debugging as shown below: 

    nextflow -trace nextflow.executor run <your script>

**Comment by pditommaso** [comment]
@dctrud Did you have any chance to test this patch ?

**Comment by dctrud** [comment]
@pditommaso I'll be testing it today and will let you know. Thanks.

**Comment by dctrud** [comment]
@pditommaso - unfortunately our cluster has been fully occupied today. I want to test there rather than just on a test VM cluster, so am having to wait for some slots so I can manually do the re-queue etc. Will let you know asap. Thanks again for your efforts.

**Comment by dctrud** [comment]
@pditommaso Doesn't quite work. When slurm re-queues into pending state it sends a SIGTERM so there is a 143 exit code that Nextflow picks up and indicates job has terminated even though the SLURM job is still there, and in pending state.

I guess for our desired behavior there has to be an option to tell nextflow to ignore an exit code, and allow the job to re-run from the start if it is re-queued and thus still exists in the SLURM scheduler. This requires ignoring / cleaning up the .exitcode exit file.

I'm guessing this is a rare requirement and we have a workaround for it now for our immediate needs. I will try to spend some time looking into Nextflow code to come up with a PR. Are the changes you made for that snapshot available to see in a git branch anywhere?

Thanks

**Test case details**

Launched main.nf with 

`NXF_VER=0.22.2-SNAPSHOT ./nextflow -trace nextflow.executor run main.nf`

Three processes - middle one sleeps for an hour. Identified job number on SLURM and did the hard requeue and hold to requeue into pending state manually:

`scontrol requeue 242652 && scontrol hold 242652`

Main.nf is: https://gist.github.com/dctrud/7f1104d77bcd86335272b7f4892c26eb

Log is: https://gist.github.com/dctrud/8356afeeaa2d884db4807c028d8e2a4f

**Comment by pditommaso** [comment]
> When slurm re-queues into pending state it sends a SIGTERM so there is a 143 exit code that Nextflow picks up and indicates job has terminated 
This is bad because that signal terminates the posix process. I was thinking a `SIGSTOP` was used instead. 

This makes things more difficult. 

 >  Are the changes you made for that snapshot available to see in a git branch anywhere?

I've just pushed [here](https://github.com/nextflow-io/nextflow/commits/fix_issue_226). If you send any PR please use this branch.

**Comment by dctrud** [comment]
> This is bad because that signal terminates the posix process. I was thinking that SIGSTOP was used instead. This makes things more difficult. 

Yes. The SIGSTOP happens when if you suspend a job and it stays on the node. The requeue is terminating the job so it vacates the node entirely and goes back to the queue.

I believe to handle requeues the logic change will be in ` nextflow/src/main/groovy/nextflow/executor/GridTaskHandler.groovy`

There is some code at line 206...
```
    // -- if the job is active, this means that it is still running and thus the exit file cannot exist
    //    returns null to continue to wait
            if( active )
                return null
```
... but this is inside an if, so the check only occurs when the exitfile doesn't exist. Putting something at the top of the function, to return null if active even before the exitfile existence is even checked may work. I guess that should be an executor option though, not a default? It's not the case that we can assume all jobs can be restarted from the top safely if they've already done some work.

I'll try some stuff out.

**Comment by pditommaso** [comment]
> Yes. The SIGSTOP happens when if you suspend a job and it stays on the node. The requeue is terminating the job so it vacates the node entirely and goes back to the queue.

This means that SLURM re-execute the job in the directory, right? I my opinion it's a very bad practice because the job when restarted can conflict with some partial result produced by the previous execution. 

> Putting something at the top of the function, to return null if active even before the exitfile existence is even checked may work.

I think the best approach would be to instruct the executor to ignore the `143` status instead of relying on the SLURM status. Obviously it should be an executor configuration option and not hard-code.

**Comment by pditommaso** [comment]
How is the status of this issue? Have you managed to solve on your side or it's still an open problem?

**Comment by dctrud** [comment]
Hi - I'm not really working on this right know, as I just had an interruption in the form of a son being born a little earlier than expected. Will pick it up when I return to work in early November.

**Comment by pditommaso** [comment]
This issue has been fixed by adding `--no-requeue` option in the SLURM job submission, included in version `0.22.4`.
